network/src/noise/stream.rs

```diff
@@ -303,39 +296,20 @@ where
                         return Poll::Ready(Ok(Some(bytes_buffered)));
                     }
                 },
-                WriteState::WriteFrameLen {
-                    frame_len,
-                    ref buf,
-                    ref mut offset,
-                } => {
-                    match ready!(poll_write_all(
-                        context,
-                        Pin::new(&mut self.socket),
-                        buf,
-                        offset
-                    )) {
-                        Ok(()) => {
-                            self.write_state = WriteState::WriteEncryptedFrame {
-                                frame_len,
-                                offset: 0,
-                            };
-                        },
-                        Err(e) => {
-                            if e.kind() == io::ErrorKind::WriteZero {
-                                self.write_state = WriteState::Eof;
-                            }
-                            return Poll::Ready(Err(e));
-                        },
-                    }
-                },
                 WriteState::WriteEncryptedFrame {
                     frame_len,
                     ref mut offset,
                 } => {
+                    // Create a buffer with the message len prepended to the message data
+                    let frame_len_bytes = &u16::to_be_bytes(frame_len);
+                    let message_bytes = &self.buffers.write_buffer[..(frame_len as usize)];
+                    let message_and_len_bytes = [frame_len_bytes, message_bytes].concat();
+
```

Comment:
this sucks... most APIs actually expect this type of behavior and expose a vector of byte arrays: https://docs.rs/tokio/latest/tokio/io/trait.AsyncWrite.html#method.poll_write_vectored

of course the gain here is probably minimal, but this does result in an extra copy... so now you've made a conscious decision that an extra copy is cheaper than an OS call. That's probably a safe bet, but I can imagine in some cases it isn't.

Reply:
> so now you've made a conscious decision that an extra copy is cheaper than an OS call. That's probably a safe bet, but I can imagine in some cases it isn't.
Yeah, agreed. I think the best solution is for the write buffer to actually allocate data + 2 bytes first, and then we write the length to the first 2 bytes after we know the encrypted length. But, I'll leave it as a follow up (i.e., baby steps 😄) 

> most APIs actually expect this type of behavior and expose a vector of byte arrays
Yeah, I briefly looked into this but the problem that I saw was that the default implementation of `poll_write_vectored` simply calls `poll_write` on the first non-empty buffer. So, even if I gave both buffers to `poll_write_vectored`, it would still make two `poll_write` calls, which seems odd to me. (See: https://rcos.io/static/internal_docs/src/futures_io/lib.rs.html#158). The comment mentions `Objects which support vectored IO should override this method.` -- but I'm not sure if that actually happens 🤔